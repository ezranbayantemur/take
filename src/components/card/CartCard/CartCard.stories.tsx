import type {Meta, StoryObj} from '@storybook/react-native';
import {CartCard} from '@components';
import {Product} from '@types';
import type {CartCardProps} from './CartCard.types';

import productsData from '@mocks/products_for_category.json';

const mockData: Product = productsData[2];

const CartCardMeta: Meta<CartCardProps> = {
  args: {
    product: mockData,
    quantity: 3,
  },
  title: 'CartCard',
  component: CartCard,
};

export default CartCardMeta;

export const Basic: StoryObj<CartCardProps> = {};
