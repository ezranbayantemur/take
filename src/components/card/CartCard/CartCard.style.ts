import {StyleSheet} from 'react-native';
import {color, font, padding, typography} from '@style';

export default StyleSheet.create({
  container: {
    margin: padding.SMALL,
    borderColor: color.SMOKE,
    flexDirection: 'row',
  },
  image: {
    margin: padding.SMALL,
    width: 100,
    height: 100,
  },
  name: {
    fontFamily: font.BOLD,
    fontSize: typography.NORMAL,
    textAlign: 'center',
    paddingBottom: padding.SMALL,
  },
  quantity: {
    fontFamily: font.REGULAR,
    fontSize: typography.SMALL,
  },
  price: {
    fontFamily: font.BOLD_ITALIC,
  },
  product_info_container: {
    flex: 1,
    paddingLeft: padding.SMALL,
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  price_container: {
    flex: 1,
    padding: padding.TINY,
    alignItems: 'flex-end',
    justifyContent: 'center',
  },
  total_price: {
    position: 'absolute',
    bottom: 5,
    right: 5,
    fontSize: typography.NORMAL,
    fontFamily: font.BLACK,
  },
  total_title: {
    fontFamily: font.REGULAR,
  },
  close_button_icon: {
    marginTop: padding.SMALL,
    alignSelf: 'flex-start',
  },
});
