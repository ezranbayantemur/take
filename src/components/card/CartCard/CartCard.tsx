import {Image, Text, View} from 'react-native';
import React from 'react';
import InputCounter from '../../InputCounter';
import styles from './CartCard.style';
import type {CartCardProps} from './CartCard.types';

const CartCard = ({
  testID,
  product,
  quantity,
  onRemoveFromCart,
  onIncreaseQuantity,
  onDecreaseQuantity,
}: CartCardProps) => {
  const totalPrice = (product.product_price * quantity).toLocaleString('tr-TR');

  return (
    <View testID={`${testID}_cartcard`} style={styles.container}>
      <Image
        resizeMode="contain"
        style={styles.image}
        source={{
          uri: product.product_image,
        }}
      />
      <View style={styles.product_info_container}>
        <View>
          <Text style={styles.name}>{product.product_name}</Text>
          <Text style={styles.quantity}>Adet: {quantity}</Text>
          <Text style={styles.price}>{product.product_price} TL</Text>
        </View>
      </View>
      <View style={styles.price_container}>
        <InputCounter
          testID={`${testID}_cartcard_inputcounter`}
          value={quantity}
          onIncrease={onIncreaseQuantity}
          onDecrease={onDecreaseQuantity}
        />
        <Text style={styles.total_price}>
          <Text style={styles.total_title}>Toplam:</Text> {totalPrice} TL
        </Text>
      </View>
    </View>
  );
};

export default React.memo(CartCard);
