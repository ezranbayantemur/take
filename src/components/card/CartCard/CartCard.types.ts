import type {Product} from '@types';

export interface CartCardProps {
  testID: string;
  product: Product;
  quantity: number;
  onRemoveFromCart: () => void;
  onDecreaseQuantity: () => void;
  onIncreaseQuantity: () => void;
}
