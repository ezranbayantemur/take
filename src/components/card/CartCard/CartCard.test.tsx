import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import CartCard from './CartCard';
import productsData from '@mocks/products_for_category.json';
import {Product} from '@types';

const mockData: Product = productsData[2];
const mockOnRemove = jest.fn();
const mockOnDecreaseQuantity = jest.fn();
const mockOnIncreaseQuantity = jest.fn();

let wrapper: ReturnType<typeof render>;

// TODO: Add tests
describe('CartCard unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <CartCard
        testID="test"
        product={mockData}
        quantity={3}
        onRemove={mockOnRemove}
        onDecreaseQuantity={mockOnDecreaseQuantity}
        onIncreaseQuantity={mockOnIncreaseQuantity}
      />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render titles correctly', () => {
    expect(wrapper.queryByText('ThinkPhone')).not.toBeNull();
    expect(wrapper.queryByText('Adet: 3')).not.toBeNull();
    expect(wrapper.queryByText('150 TL')).not.toBeNull();
    expect(wrapper.queryByText('Toplam: 450 TL')).not.toBeNull();
  });

  // Gives error: Unable to find node on an unmounted component.
  it.skip('should update total price and quantity title correctly if quantity increase', async () => {
    const _wrapper = render(
      <CartCard
        testID="test"
        product={mockData}
        quantity={3}
        onRemove={mockOnRemove}
        onDecreaseQuantity={mockOnDecreaseQuantity}
        onIncreaseQuantity={mockOnIncreaseQuantity}
      />,
    );

    const increaseButton = _wrapper.getByTestId(
      'test_cartcard_inputcounter_increase_touchable',
    );
    expect(_wrapper.queryByText('Adet: 3')).not.toBeNull();
    expect(_wrapper.queryByText('Toplam: 450 TL')).not.toBeNull();

    fireEvent.press(increaseButton);

    waitFor(() => {
      expect(wrapper.queryByText('Adet: 4')).not.toBeNull();
      expect(wrapper.queryByText('Toplam: 600 TL')).not.toBeNull();
    });
  });

  // Gives error: Unable to find node on an unmounted component.
  it.skip('should update total price and quantity title correctly if quantity decrease', () => {
    const decreaseButton = wrapper.getByTestId(
      'test_cartcard_inputcounter_decrease_touchable',
    );
    expect(wrapper.queryByText('Adet: 3')).not.toBeNull();
    expect(wrapper.queryByText('Toplam: 450 TL')).not.toBeNull();

    fireEvent(decreaseButton, 'onPress');

    waitFor(() => {
      expect(wrapper.queryByText('Adet: 2')).not.toBeNull();
      expect(wrapper.queryByText('Toplam: 300 TL')).not.toBeNull();
    });
  });

  it('should trigger remove', () => {
    const remove = wrapper.getByTestId('test_cartcard_remove');
    fireEvent(remove, 'onPress');

    expect(mockOnRemove).toBeCalled();
  });
});
