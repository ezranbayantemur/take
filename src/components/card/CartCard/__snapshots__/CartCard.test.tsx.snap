// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartCard unit tests should render correctly 1`] = `
<View
  style={
    {
      "borderColor": "#BDBDBD",
      "flexDirection": "row",
      "margin": 8,
    }
  }
  testID="test_cartcard"
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "https://motorolaus.vtexassets.com/arquivos/ids/162812/thinkphone-pdp-ecom-render-1-carbon-black-n5re8wvk.png?v=638084623518870000",
      }
    }
    style={
      {
        "height": 100,
        "margin": 8,
        "width": 100,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingLeft": 8,
      }
    }
  >
    <View>
      <Text
        style={
          {
            "fontFamily": "Lato-Bold",
            "fontSize": 15,
          }
        }
      >
        ThinkPhone
      </Text>
      <Text
        style={
          {
            "fontFamily": "Lato-Regular",
            "fontSize": 13,
          }
        }
      >
        Adet: 
      </Text>
      <Text
        style={
          {
            "fontFamily": "Lato-BoldItalic",
          }
        }
      >
        17.500
         TL
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flex": 1,
        "justifyContent": "center",
        "padding": 4,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#DB040E",
            "borderRadius": 20,
            "height": 25,
            "justifyContent": "center",
            "marginHorizontal": 8,
            "opacity": 1,
            "width": 25,
          }
        }
        testID="test_cartcard_inputcounter_increase_touchable"
      >
        <Text
          style={
            {
              "color": "#FBFBFB",
              "fontFamily": "Lato-Black",
              "fontSize": 15,
            }
          }
        >
          +
        </Text>
      </View>
      <Text
        style={
          {
            "fontFamily": "Lato-Black",
            "fontSize": 17,
          }
        }
        testID="test_cartcard_inputcounter_value_text"
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#DB040E",
            "borderRadius": 20,
            "height": 25,
            "justifyContent": "center",
            "marginHorizontal": 8,
            "opacity": 1,
            "width": 25,
          }
        }
        testID="test_cartcard_inputcounter_decrease_touchable"
      >
        <Text
          style={
            {
              "color": "#FBFBFB",
              "fontFamily": "Lato-Black",
              "fontSize": 15,
            }
          }
        >
          -
        </Text>
      </View>
    </View>
    <Text
      style={
        {
          "bottom": 5,
          "fontFamily": "Lato-Black",
          "fontSize": 15,
          "position": "absolute",
          "right": 5,
        }
      }
    >
      <Text
        style={
          {
            "fontFamily": "Lato-Regular",
          }
        }
      >
        Toplam:
      </Text>
       
      NaN
       TL
    </Text>
  </View>
</View>
`;
